File: c++/parser.cpp
#include <iostream>
#include <memory>

#include "doctest.h"
#include "lexer.h"
#include "parser.h"

void Parser::parse() {
    for (auto token : _tokens) {
        std::cout << "Token: " << token.value() << std::endl;
    }
    /* Program ast; */
    /* return ast; */
}

void Parser::expect(std::string expected) {
    const Token &actual = _tokens[_current_token++];
    if (actual.value() != expected) {
        throw SyntaxError(
            std::format("Expected {}, but got {}", expected, actual.value()));
    }
}

std::unique_ptr<Statement> Parser::parse_statement() {
    expect("return");
    auto exp = parse_exp();
    expect(";");
    return std::make_unique<Return>(std::move(exp));
}

std::unique_ptr<Exp> Parser::parse_exp() {
    const Token &token = _tokens[_current_token++];
    return std::make_unique<Constant>(token.value());
}

TEST_CASE("Parser::parse_statement") {
    std::vector<Token> tokens{{"return"}, {"1234"}, {";"}};
    Parser parser(tokens);
    auto stmt = parser.parse_statement();
    CHECK(stmt->to_string() == "Return(\n  Constant(1234)\n)");
}

TEST_CASE("Parser::parse_exp") {
    Parser parser(std::vector<Token>({{"100"}}));
    auto exp = parser.parse_exp();
    CHECK(exp->to_string() == "Constant(100)");
}

TEST_CASE("expressions have string representations") {
    Constant exp{"42"};
    CHECK(exp.to_string() == "Constant(42)");
}

TEST_CASE("statements have string representations") {
    Return stmt(std::make_unique<Constant>("23"));
    CHECK(stmt.to_string() == "Return(\n  Constant(23)\n)");
}

File: rust/src/parser.rs
use crate::lexer::Token;
use core::fmt;

pub enum Exp {
    Constant(String),
}

impl fmt::Display for Exp {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Exp::Constant(v) => write!(f, "Constant({})", v),
        }
    }
}

enum Statement {
    Return(Exp),
}

impl fmt::Display for Statement {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Statement::Return(exp) => write!(f, "Return({})", exp),
        }
    }
}

pub struct Parser<'a> {
    tokens: &'a Vec<Token>,
    current_token: usize,
}

impl<'a> Parser<'a> {
    pub fn new(tokens: &'a Vec<Token>) -> Self {
        Self {
            tokens,
            current_token: 0,
        }
    }

    pub fn parse(&self) {}

    pub fn parse_exp(&mut self) -> Exp {
        let token = &self.tokens[self.current_token];
        self.current_token += 1;
        match token {
            Token::Constant(v) => Exp::Constant(v.clone()),
            t => panic!("Not a constant: {:?}", t),
        }
    }

    fn parse_statement(&mut self) -> Statement {
        self.expect(Token::Identifier("return".to_string()));
        let exp = self.parse_exp();
        self.expect(Token::Literal(";".to_string()));
        Statement::Return(exp)
    }

    fn expect(&mut self, expected: Token) {
        if self.current_token == self.tokens.len() {
            panic!("Missing \"{:?}\"", expected);
        }
        let actual = &self.tokens[self.current_token];
        self.current_token += 1;
        if actual != &expected {
            panic!("Expected \"{:?}\" but got \"{:?}\"", expected, actual);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    #[should_panic]
    fn parse_statement_with_missing_expression() {
        // TODO: `return;` is actually valid C
        let tokens = vec![ident("return"), lit(";")];
        let mut parser = Parser::new(&tokens);
        parser.parse_statement();
    }

    #[test]
    #[should_panic]
    fn parse_statement_with_missing_semicolon() {
        let tokens = vec![ident("bork"), constant("1234")];
        let mut parser = Parser::new(&tokens);
        parser.parse_statement();
    }

    #[test]
    #[should_panic]
    fn parse_statement_with_missing_identifier() {
        let tokens = vec![ident("bork"), constant("1234"), lit(";")];
        let mut parser = Parser::new(&tokens);
        parser.parse_statement();
    }

    #[test]
    fn parse_statement_success() {
        let tokens = vec![ident("return"), constant("1234"), lit(";")];
        let mut parser = Parser::new(&tokens);
        let stmt = parser.parse_statement();
        assert_eq!(stmt.to_string(), "Return(Constant(1234))");
    }

    #[test]
    #[should_panic]
    fn parse_exp_with_error() {
        let tokens = vec![Token::Identifier("bark".to_string())];
        let mut parser = Parser::new(&tokens);
        parser.parse_exp();
    }

    #[test]
    fn parse_exp_with_constant() {
        let tokens = vec![Token::Constant("100".to_string())];
        let mut parser = Parser::new(&tokens);
        let exp = parser.parse_exp();
        assert_eq!(exp.to_string(), "Constant(100)");
    }

    fn ident(value: &str) -> Token {
        Token::Identifier(value.to_string())
    }

    fn constant(value: &str) -> Token {
        Token::Constant(value.to_string())
    }

    fn lit(value: &str) -> Token {
        Token::Literal(value.to_string())
    }
}
